python_benchmarks/core.py

```diff
@@ -0,0 +1,202 @@
+import ctypes
+import torch
+from torch.profiler import profile, ProfilerActivity
+from typing import List, Callable, Union
+from torch.autograd import DeviceType
+
+
+def get_device_properties() -> tuple:
+    """
+    Computes L2 cache size and peak device bandwidth using ctypes and cuda.
+    """
+    libnames = ("libcuda.so", "libcuda.dylib", "nvcuda.dll", "cuda.dll")
+    for libname in libnames:
+        try:
+            cuda = ctypes.CDLL(libname)
+        except OSError:
+            continue
+        else:
+            break
+    else:
+        raise OSError("could not load any of: " + " ".join(libnames))
+
+    # Device attribute enums (taken from cuda.h)
+    # https://nvidia.github.io/cuda-python/module/cuda.html
+
+    CU_DEVICE_ATTRIBUTE_MEMORY_CLOCK_RATE = 36
+    CU_DEVICE_ATTRIBUTE_GLOBAL_MEMORY_BUS_WIDTH = 37
+    CU_DEVICE_ATTRIBUTE_L2_CACHE_SIZE = 38
+
+    l2_cache_size = ctypes.c_int()
+    device = torch.cuda.current_device()
+    memory_clock_rate = ctypes.c_int()
+    memory_bus_width = ctypes.c_int()
+
+    cuda.cuDeviceGetAttribute(
+        ctypes.byref(l2_cache_size), CU_DEVICE_ATTRIBUTE_L2_CACHE_SIZE, device
+    )
+    cuda.cuDeviceGetAttribute(
+        ctypes.byref(memory_clock_rate), CU_DEVICE_ATTRIBUTE_MEMORY_CLOCK_RATE, device
+    )
+    cuda.cuDeviceGetAttribute(
+        ctypes.byref(memory_bus_width),
+        CU_DEVICE_ATTRIBUTE_GLOBAL_MEMORY_BUS_WIDTH,
+        device,
+    )
+
+    # Compute peak bandwidth in GBps
+    peak_bandwidth = (2 * memory_bus_width.value * memory_clock_rate.value) / (1e6 * 8)
+
+    return l2_cache_size.value, peak_bandwidth
+
+
+L2_CACHE_SIZE, PEAK_BANDWIDTH = get_device_properties()
```

Comment:
I'd rename `PEAK_BANDWIDTH` to `PEAK_BANDWIDTH_GBPS` so the unit is clear when it's used. (Ideally, I'd rename `L2_CACHE_SIZE` as well but I'm fine with the convention `size` is by default in bytes).